Fix #1580 Axis renderers do not render all tick values they are provided

Jonarw opened the pull request:
Fixes #1580.

### Checklist

- [x] I have included examples or tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file
- [x] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- Remove tick value checks from the axis renderers
- Change `AxisUtilities.CreateTickValues(...)` such that it only returns tick values inside the requested range
- Add `AxisUtilities.FilterRedundantMinorTicks(...)` which can be used to filter out minor ticks that are not needed because they are 'on top of' a major tick

Because these changes affect pretty much all kinds of plots, there is a lot of possibility of breaking things. Therefore I would be thankful if someone could second-guess this and tell me what I have forgotten...

@oxyplot/admins

VisualMelon commented:
I got an exception with Polar Plots > Semi-circle full plot area. I'm afraid I have no idea what caused it and can't seem to reproduce it, but I was resizing the example browser window at the time.

![image](https://user-images.githubusercontent.com/5646475/83275743-8333df00-a1c7-11ea-9f8e-4aa3d6134485.png)

It happened while I was trying to get a screen shot of the magnitude axis rendering labels beyond the plot bounds:

![image](https://user-images.githubusercontent.com/5646475/83276152-19680500-a1c8-11ea-8921-ca0111c8b5ca.png)

This behaviour isn't new; I don't know whether the crash is. I wouldn't worry about it for the moment, but maybe message will help.

<hr />

I have tried to break the X/Y Axes, but they just seem to be improved over the old ones (minor tick filtering is more reliable).

VisualMelon commented on Source/OxyPlot/Axes/Rendering/MagnitudeAxisFullPlotAreaRenderer.cs:
```diff
@@ -115,7 +114,7 @@ public override void Render(Axis axis, int pass)
             {
                 OxyPen pen = this.MinorPen;
                 //var minorTicks = this.MinorTickValues.Where(x => x >= axis.ActualMinimum && x <= axis.ActualMaximum && !majorTicks.Contains(x)).ToArray();
```
This comment could be removed.

VisualMelon commented:
There is a regression with `MagnitudeAxisFullPlotArea`: pre-filtering the minor tick values means that `ExtendTicks` can go wrong:

![image](https://user-images.githubusercontent.com/5646475/83283297-e4ad7b00-a1d2-11ea-9cca-68165ed75335.png)

The outer-most minor ticks are not the right distance apart, so the 'extended' ticks are too spaced out. The example shown is `ArchimedeanSpiralFullPlotArea` with the series changed to:

    model.Series.Add(new FunctionSeries(t => t, t => t, 0, Math.PI * 6, 0.01));

No existing example presents this issue.

Jonarw commented:
Good find, thanks!
That `MagnitudeAxisFullPlotAreaRenderer` is a bit of a mess...

I've added a commit where I have moved the responsibility for determining the ticks to display out of the `MagnitudeAxisFullPlotAreaRenderer` and into `MagnitudeAxisFullPlotArea` to be consistent with the rest of this PR. This eliminates the need for the `ExtendTicks` method altogether, which is a good thing in my opinion.

VisualMelon commented on Source/OxyPlot/Axes/AxisUtilities.cs:
```diff
@@ -106,8 +106,18 @@ public static IList<double> FilterRedundantMinorTicks(IList<double> majorTicks,
 
+            static double GetEpsilon(double tick1, double tick2)
+            {
+                return Math.Abs(tick1 - tick2) * 1e-3;
```
I wonder if this could have issues with large values for `LogarithmicAxis.Base` (e.g. on the order of 10/1e-3).

Jonarw replied:
Interesting thought!
I don't really know of an application where such large bases would be used, but of course that doesn't mean at all that they don't exist.

Just for fun I modified the LogarithmicAxis>Tick Calculation example by multiplying the bases by 1000. The result is kind of 'interesting':
![image](https://user-images.githubusercontent.com/18097859/83324373-2129b700-a265-11ea-8584-284460909eab.png)

So yes I do think `AxisUtilities.FilterRedundantMinorTicks(...)` could have a problem with these kinds of bases, but I don't think we really support them anyway.

VisualMelon replied:
That first image looks like a bug in computing the minor tick interval rather than your stuff: did you happen to see what happens with the old code?

I agree this shouldn't be a blocker either way.

Jonarw replied:
The first image is because the `LogarithmicAxis` falls back to linearly-spaced ticks when the range is too small. That's what I meant with we don't support these kinds of bases anyway.

Before this PR I reckon this would have looked pretty much the same.

Jonarw commented:
A bit of an oddity is that we see on many axis labels the tick label '-0' with this PR, for example:
![image](https://user-images.githubusercontent.com/18097859/83325684-6b636600-a26e-11ea-9984-662b06adc87d.png)

I think I understand why this happens, this should just be because we display 5 significant digits from something like -0.0000000001. I am unsure however how to best fix this. I don't think the tick calculation does anything wrong here, this is just what you get when you work with floating-point numbers. 

I think this should be addressed in the label formatter. Could we just hard-code the case '-0' and return '0' instead? Is there a better way? Are there other edge cases like this?

VisualMelon commented:
I'd guess it is because `Math.Ceiling(-0.0000000001)` gives you `-0` instead of `0`. This could be the place to 'fix' this, I think.

Jonarw commented:
Oh yes you are right of course!

Jonarw commented:
(squashed)
I think I am pretty happy with this now, unless you can find some more issues I've overlooked?

Jonarw commented:
Apparently on AppVeyor a test of some Histogram example just randomly failed...

VisualMelon commented:
@Jonarw that's my fault: looks like a _random_ example failed. That can be ignored; I'll make the example deterministic some time in the future.

Jonarw commented:
(rebased)

Jonarw commented:
From my side this would be ready to merge.
Any more concerns?

VisualMelon commented:
@Jonarw if it's OK with you, I'd like to take another look (can probably do it tonight if I don't get distracted; otherwise, pester me tomorrow). I have also not looked at the latest changes to `MagnitudeAxisFullPlotAreaRenderer` at all yet.

VisualMelon commented:
Everything here looks line to me. `MagnitudeAxisFullPlotArea` looks much tidier now.

Jonarw commented:
Thanks for taking the time!
